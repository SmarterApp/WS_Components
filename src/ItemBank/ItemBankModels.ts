import {
  ItemModel,
  AboutItemRevisionModel,
  GradeLevel,
  RevisionModel
} from "@src/index";

export interface ItemRevisionModel {
  itemKey?: number;
  bankKey?: number;
  section?: string;
  revision?: string;
  isaap?: string;
}

export function getItemBankName(itemRevisionModel: ItemRevisionModel) {
  let value = "";
  if (itemRevisionModel.bankKey && itemRevisionModel.itemKey) {
    value = `${itemRevisionModel.bankKey}-${itemRevisionModel.itemKey}`;
  }

  return value;
}

export function itemRevisionKey(itemRevisionModel: ItemRevisionModel) {
  return `${itemRevisionModel.bankKey}-${itemRevisionModel.itemKey}-${
    itemRevisionModel.section
  }`;
}

export function validItemRevisionModel(itemRevisionModel: ItemRevisionModel) {
  let value = false;
  if (
    itemRevisionModel.itemKey &&
    itemRevisionModel.bankKey &&
    itemRevisionModel.section
  ) {
    value = true;
  }

  return value;
}

export function getItemBankIndex(
  currentItem: ItemRevisionModel,
  items: ItemRevisionModel[]
) {
  const key = itemRevisionKey(currentItem);

  return items.findIndex(it => itemRevisionKey(it) === key);
}

export function getNextItemBank(
  currentItem: ItemRevisionModel,
  items: ItemRevisionModel[]
): ItemRevisionModel | undefined {
  const currentIdx = getItemBankIndex(currentItem, items);
  let nextItem: ItemRevisionModel | undefined;

  const nextIdx = currentIdx + 1;
  if (
    nextIdx >= 0 &&
    nextIdx < items.length &&
    validItemRevisionModel(items[nextIdx])
  ) {
    nextItem = items[nextIdx];
  }

  return nextItem;
}

export function getPreviousItemBank(
  currentItem: ItemRevisionModel,
  items: ItemRevisionModel[]
): ItemRevisionModel | undefined {
  let previousItem: ItemRevisionModel | undefined;
  const currentIdx = getItemBankIndex(currentItem, items);

  const prevIdx = currentIdx - 1;
  if (prevIdx >= 0 && validItemRevisionModel(items[prevIdx])) {
    previousItem = items[prevIdx];
  }

  return previousItem;
}

export interface SectionModel {
  key: string;
  value: string;
}

export interface AccessibilityRevisionModel {
  gradeLevel: GradeLevel;
  subject: string;
  interactionType: string;
  allowCalculator?: boolean;
  isPerformance?: boolean;
}
