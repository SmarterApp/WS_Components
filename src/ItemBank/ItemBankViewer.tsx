import * as React from "react";
import {
  AboutItemRevisionModel,
  AccResourceGroupModel,
  AdvancedAboutItem,
  ItemAccessibilityModal,
  ItemViewerFrame,
  ResourceSelectionsModel,
  RevisionContainer,
  Revision,
  ItemRevisionModel,
  itemRevisionKey,
  getItemBankName,
  RevisionModel,
  RubricModal
} from "../index";

export interface ItemBankViewerProps {
  onAccessibilityUpdate: (accResourceGroups: AccResourceGroupModel[]) => void;
  onAccessibilityReset: () => void;
  onItemSelect: (direction: "next" | "previous") => void;
  onRevisionSelect: (revision: string) => void;
  itemUrl?: string;
  aboutItemRevisionModel?: AboutItemRevisionModel;
  accResourceGroups?: AccResourceGroupModel[];
  revisions?: RevisionModel[];
  nextItem?: ItemRevisionModel;
  prevItem?: ItemRevisionModel;
}

export class ItemBankViewer extends React.Component<ItemBankViewerProps, {}> {
  constructor(props: ItemBankViewerProps) {
    super(props);
  }

  renderRightNav(): JSX.Element | undefined {
    const {
      onAccessibilityUpdate,
      onAccessibilityReset,
      accResourceGroups
    } = this.props;

    let content: JSX.Element | undefined;
    if (accResourceGroups) {
      content = (
        <div
          className="item-nav-right-group"
          role="group"
          aria-label="Second group"
        >
          <ItemAccessibilityModal
            accResourceGroups={accResourceGroups}
            onSave={onAccessibilityUpdate}
            onReset={onAccessibilityReset}
          />
        </div>
      );
    }

    return content;
  }

  renderNavButton(direction: "next" | "previous") {
    const { onItemSelect, nextItem, prevItem } = this.props;
    let label = "";
    let itemName = "";
    let item = nextItem;
    if (direction === "previous") {
      itemName = prevItem ? getItemBankName(prevItem) : " ";
      label = "Previous";
      item = prevItem;
    } else {
      itemName = nextItem ? getItemBankName(nextItem) : " ";
      label = "Next";
    }

    return (
      <button
        onClick={() => onItemSelect(direction)}
        className={"btn btn-link " + direction}
        disabled={item ? false : true}
      >
        <div className="nav-button-container">
          <span
            className={
              direction === "previous"
                ? "fa fa-arrow-left"
                : "fa fa-arrow-right"
            }
            aria-hidden="true"
          />
          <div className="nav-item-link">
            <div>{label}</div>
            <div className="nav-item-name">{itemName}</div>
          </div>
        </div>
      </button>
    );
  }

  renderMidNav() {
    const { onItemSelect, nextItem, prevItem } = this.props;
    const nextItemName = nextItem ? getItemBankName(nextItem) : " ";
    const previousItemName = prevItem ? getItemBankName(prevItem) : " ";

    return (
      <div className="nav-buttons">
        {this.renderNavButton("previous")}
        {this.renderNavButton("next")}
      </div>
    );
  }

  renderRubricModal(): JSX.Element | undefined {
    const { aboutItemRevisionModel } = this.props;
    let content: JSX.Element | undefined;
    if (
      aboutItemRevisionModel &&
      aboutItemRevisionModel.sampleItemScoring &&
      aboutItemRevisionModel.sampleItemScoring.rubrics
    ) {
      const rubrics = aboutItemRevisionModel.sampleItemScoring.rubrics;
      content = <RubricModal rubrics={rubrics} />;
    }

    return content;
  }

  renderNavBar(): JSX.Element | undefined {
    const { aboutItemRevisionModel, accResourceGroups } = this.props;
    let content: JSX.Element | undefined;
    if (aboutItemRevisionModel && accResourceGroups) {
      content = (
        <div
          className="item-nav"
          role="toolbar"
          aria-label="Toolbar with button groups"
        >
          <div
            className="item-nav-left-group"
            role="group"
            aria-label="First group"
          >
            <AdvancedAboutItem {...aboutItemRevisionModel} />
            {this.renderRubricModal()}
          </div>
          {this.renderMidNav()}
          {this.renderRightNav()}
        </div>
      );
    }

    return content;
  }

  render() {
    const { itemUrl, revisions, onRevisionSelect } = this.props;

    return (
      <div className="section revisions-items">
        <RevisionContainer
          revisions={revisions || []}
          onRevisionSelect={onRevisionSelect}
        />
        <div
          className="section item-map-iframe"
          aria-live="polite"
          aria-relevant="additions removals"
        >
          {this.renderNavBar()}
          <ItemViewerFrame url={itemUrl} title={"item viewer"} />
        </div>
      </div>
    );
  }
}
