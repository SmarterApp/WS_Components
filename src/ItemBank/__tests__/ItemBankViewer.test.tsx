import * as React from "react";
import { shallow } from "enzyme";
import { ItemBankViewer } from "../ItemBankViewer";
import {
  AboutItemRevisionModel,
  AccResourceGroupModel,
  RevisionModel
} from "../../index";
import { ItemRevisionModel } from "../ItemBankModels";

const date = new Date("2015-11-16T00:00:00");

const revision: RevisionModel = {
  author: "Troy",
  date: date,
  commitMessage: "test",
  commitHash: "asdh49"
};

const aboutItemRevisionModel: AboutItemRevisionModel = {
  itemKey: "187",
  bankKey: "3000",
  revision: "hello",
  section: "math",
  AboutItemMetadata: undefined,
  sampleItemScoring: undefined
};

const accResourceGroup: AccResourceGroupModel = {
  label: "none",
  order: 1,
  accessibilityResources: []
};

const accResourceGroups: AccResourceGroupModel[] = [accResourceGroup];

const revisions: RevisionModel[] = [revision];

const nextItem: ItemRevisionModel = {
  itemKey: 187,
  bankKey: 3000,
  section: "math",
  revision: "ajweio",
  isaap: undefined
};

const prevItem: ItemRevisionModel = {
  itemKey: 187,
  bankKey: 3000,
  section: "math",
  revision: "ajweio",
  isaap: undefined
};

describe("ItemBankViewer", () => {
  const wrapper = shallow(
    <ItemBankViewer
      onAccessibilityReset={() => {}}
      onAccessibilityUpdate={() => {}}
      onItemSelect={() => {}}
      onRevisionSelect={() => {}}
    />
  );

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe("ItemBankViewer with revision model", () => {
  const wrapper = shallow(
    <ItemBankViewer
      onAccessibilityReset={() => {}}
      onAccessibilityUpdate={() => {}}
      onItemSelect={() => {}}
      onRevisionSelect={() => {}}
      aboutItemRevisionModel={aboutItemRevisionModel}
    />
  );
  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe("ItemBankViewer with revision model and accResourceGroup", () => {
  const wrapper = shallow(
    <ItemBankViewer
      onAccessibilityReset={() => {}}
      onAccessibilityUpdate={() => {}}
      onItemSelect={() => {}}
      onRevisionSelect={() => {}}
      aboutItemRevisionModel={aboutItemRevisionModel}
      accResourceGroups={accResourceGroups}
    />
  );
  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe("ItemBankViewer with all props defined", () => {
  const wrapper = shallow(
    <ItemBankViewer
      onAccessibilityReset={() => {}}
      onAccessibilityUpdate={() => {}}
      onItemSelect={() => {}}
      onRevisionSelect={() => {}}
      aboutItemRevisionModel={aboutItemRevisionModel}
      accResourceGroups={accResourceGroups}
      revisions={revisions}
      nextItem={nextItem}
      prevItem={prevItem}
    />
  );
  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
