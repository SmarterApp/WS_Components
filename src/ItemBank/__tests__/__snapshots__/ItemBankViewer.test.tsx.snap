// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemBankViewer matches snapshot 1`] = `
<div
  className="section revisions-items"
>
  <RevisionContainer
    onRevisionSelect={[Function]}
    revisions={Array []}
  />
  <div
    aria-live="polite"
    aria-relevant="additions removals"
    className="section item-map-iframe"
  >
    <ItemViewerFrame
      title="item viewer"
    />
  </div>
</div>
`;

exports[`ItemBankViewer with all props defined matches snapshot 1`] = `
<div
  className="section revisions-items"
>
  <RevisionContainer
    onRevisionSelect={[Function]}
    revisions={
      Array [
        Object {
          "author": "Troy",
          "commitHash": "asdh49",
          "commitMessage": "test",
          "date": 2018-03-06T21:16:15.383Z,
        },
      ]
    }
  />
  <div
    aria-live="polite"
    aria-relevant="additions removals"
    className="section item-map-iframe"
  >
    <div
      aria-label="Toolbar with button groups"
      className="item-nav"
      role="toolbar"
    >
      <div
        aria-label="First group"
        className="item-nav-left-group"
        role="group"
      >
        <AdvancedAboutItem
          bankKey="3000"
          itemKey="187"
          revision="hello"
          section="math"
        />
      </div>
      <div
        className="nav-buttons"
      >
        <button
          className="btn btn-link previous"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="nav-button-container"
          >
            <span
              aria-hidden="true"
              className="fa fa-arrow-left"
            />
            <div
              className="nav-item-link"
            >
              <div>
                Previous
              </div>
              <div
                className="nav-item-name"
              >
                3000-187
              </div>
            </div>
          </div>
        </button>
        <button
          className="btn btn-link next"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="nav-button-container"
          >
            <span
              aria-hidden="true"
              className="fa fa-arrow-right"
            />
            <div
              className="nav-item-link"
            >
              <div>
                Next
              </div>
              <div
                className="nav-item-name"
              >
                3000-187
              </div>
            </div>
          </div>
        </button>
      </div>
      <div
        aria-label="Second group"
        className="item-nav-right-group"
        role="group"
      >
        <ItemAccessibilityModal
          accResourceGroups={
            Array [
              Object {
                "accessibilityResources": Array [],
                "label": "none",
                "order": 1,
              },
            ]
          }
          onReset={[Function]}
          onSave={[Function]}
        />
      </div>
    </div>
    <ItemViewerFrame
      title="item viewer"
    />
  </div>
</div>
`;

exports[`ItemBankViewer with revision model and accResourceGroup matches snapshot 1`] = `
<div
  className="section revisions-items"
>
  <RevisionContainer
    onRevisionSelect={[Function]}
    revisions={Array []}
  />
  <div
    aria-live="polite"
    aria-relevant="additions removals"
    className="section item-map-iframe"
  >
    <div
      aria-label="Toolbar with button groups"
      className="item-nav"
      role="toolbar"
    >
      <div
        aria-label="First group"
        className="item-nav-left-group"
        role="group"
      >
        <AdvancedAboutItem
          bankKey="3000"
          itemKey="187"
          revision="hello"
          section="math"
        />
      </div>
      <div
        className="nav-buttons"
      >
        <button
          className="btn btn-link previous"
          disabled={true}
          onClick={[Function]}
        >
          <div
            className="nav-button-container"
          >
            <span
              aria-hidden="true"
              className="fa fa-arrow-left"
            />
            <div
              className="nav-item-link"
            >
              <div>
                Previous
              </div>
              <div
                className="nav-item-name"
              >
                 
              </div>
            </div>
          </div>
        </button>
        <button
          className="btn btn-link next"
          disabled={true}
          onClick={[Function]}
        >
          <div
            className="nav-button-container"
          >
            <span
              aria-hidden="true"
              className="fa fa-arrow-right"
            />
            <div
              className="nav-item-link"
            >
              <div>
                Next
              </div>
              <div
                className="nav-item-name"
              >
                 
              </div>
            </div>
          </div>
        </button>
      </div>
      <div
        aria-label="Second group"
        className="item-nav-right-group"
        role="group"
      >
        <ItemAccessibilityModal
          accResourceGroups={
            Array [
              Object {
                "accessibilityResources": Array [],
                "label": "none",
                "order": 1,
              },
            ]
          }
          onReset={[Function]}
          onSave={[Function]}
        />
      </div>
    </div>
    <ItemViewerFrame
      title="item viewer"
    />
  </div>
</div>
`;

exports[`ItemBankViewer with revision model matches snapshot 1`] = `
<div
  className="section revisions-items"
>
  <RevisionContainer
    onRevisionSelect={[Function]}
    revisions={Array []}
  />
  <div
    aria-live="polite"
    aria-relevant="additions removals"
    className="section item-map-iframe"
  >
    <ItemViewerFrame
      title="item viewer"
    />
  </div>
</div>
`;
